server/live_query/redis_live_query.go

```diff
@@ -290,16 +337,103 @@ func (r *redisLiveQuery) QueryCompletedByHost(name string, hostID uint) error {
+	targetKey, sqlKey := generateKeys(name)
+	targets := mapBitfield(hostIDs)
+
+	// Ensure to set SQL first or else we can end up in a weird state in which a
+	// client reads that the query exists but cannot look up the SQL.
+	err := conn.Send("SET", sqlKey, sql, "EX", queryExpiration.Seconds())
+	if err != nil {
+		return fmt.Errorf("set sql: %w", err)
+	}
+	_, err = conn.Do("SET", targetKey, targets, "EX", queryExpiration.Seconds())
+	if err != nil {
+		return fmt.Errorf("set targets: %w", err)
+	}
+	return nil
+}
+
+func (r *redisLiveQuery) storeQueryNames(names ...string) error {
+	conn := redis.ConfigureDoer(r.pool, r.pool.Get())
+	defer conn.Close()
+
+	var args redigo.Args
+	args = args.Add(activeQueriesKey)
+	args = args.AddFlat(names)
+	_, err := conn.Do("SADD", args...)
+	return err
+}
+
+func (r *redisLiveQuery) removeQueryInfo(name string) error {
+	conn := redis.ConfigureDoer(r.pool, r.pool.Get())
+	defer conn.Close()
+
+	targetKey, sqlKey := generateKeys(name)
+	if _, err := conn.Do("DEL", targetKey, sqlKey); err != nil {
+		return fmt.Errorf("del query keys: %w", err)
+	}
 	return nil
 }
 
+func (r *redisLiveQuery) removeQueryNames(names ...string) error {
+	conn := redis.ConfigureDoer(r.pool, r.pool.Get())
+	defer conn.Close()
+
+	var args redigo.Args
+	args = args.Add(activeQueriesKey)
+	args = args.AddFlat(names)
+	_, err := conn.Do("SREM", args...)
+	return err
+}
+
+func (r *redisLiveQuery) loadActiveQueryNames() ([]string, error) {
+	conn := redis.ConfigureDoer(r.pool, r.pool.Get())
+	defer conn.Close()
+
+	names, err := redigo.Strings(conn.Do("SMEMBERS", activeQueriesKey))
+	if err != nil && err != redigo.ErrNil {
+		return nil, err
+	}
+	return names, nil
+}
+
 // mapBitfield takes the given host IDs and maps them into a bitfield compatible
 // with Redis. It is expected that the input IDs are in ascending order.
 func mapBitfield(hostIDs []uint) []byte {
 	if len(hostIDs) == 0 {
 		return []byte{}
 	}
 
+	// NOTE(mna): note that this is efficient storage if the host IDs are mostly
+	// sequential and starting from 1, e.g. as in a newly created database. If
+	// there's substantial churn in hosts (e.g. some are coming on and off) or
+	// for some reason the auto_increment had to be bumped (e.g. it increments with
+	// failed inserts, even if there's an "on duplicate" clause), then it could get
+	// quite inefficient. If the id gets, say, to 10M then the bitfield will take
+	// over 1MB, even if there are only 100K hosts - at which point it would
+	// likely become more efficient to store a set of host IDs (without any
+	// redis-internal storage optimization, that would be 100K * 4 bytes =
+	// ~380KB). This large bitfield usage of memory would even be true if there
+	// was only one host selected in the query, should that host be one of the
+	// high IDs.  Something to keep in mind if at some point we have reports of
+	// unexpectedly large redis memory usage, as that storage is repeated for
+	// each live query.
```

Comment:
I'd like to raise awareness on this approach and potential issue, let me know if we're comfortable keeping it that way for now (I understand that most queries probably target all or a significant subset of hosts?).

Reply:
Interesting. We do have some setups with a lot of turn over since the hosts are AWS instances that appear/disappear constantly. We can probably ignore this for now, but yeah, it could be an issue.